azuredevops/internal/service/git/resource_git_repository_file.go

```diff
@@ -0,0 +1,380 @@
+			RepositoryId: &repo,
+			CommitId:     item.CommitId,
+		})
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+
+		d.Set("commit_message", commit.Comment)
+
+		return nil
+	})
+}
+
+func resourceGitRepositoryFileUpdate(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+	ctx := context.Background()
+
+	repo := d.Get("repository_id").(string)
+	file := d.Get("file").(string)
+	branch := d.Get("branch").(string)
+
+	if err := checkRepositoryBranchExists(clients, repo, branch); err != nil {
+		return err
+	}
+
+	objectID, err := getLastCommitId(clients, repo, branch)
+	if err != nil {
+		return err
+	}
+
+	args, err := resourceGitRepositoryPushArgs(d, objectID, git.VersionControlChangeTypeValues.Edit)
+	if err != nil {
+		return err
+	}
+
+	if *(*args.Push.Commits)[0].Comment == fmt.Sprintf("Add %s", file) {
+		m := fmt.Sprintf("Update %s", file)
+		(*args.Push.Commits)[0].Comment = &m
+	}
+
+	_, err = clients.GitReposClient.CreatePush(ctx, *args)
+	if err != nil {
+		return err
+	}
+
+	return resourceGitRepositoryFileRead(d, m)
+}
+
+func resourceGitRepositoryFileDelete(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+	ctx := context.Background()
+
+	repo := d.Get("repository_id").(string)
+	file := d.Get("file").(string)
+	branch := d.Get("branch").(string)
+	message := fmt.Sprintf("Delete %s", file)
+
+	objectID, err := getLastCommitId(clients, repo, branch)
+	if err != nil {
+		return err
+	}
+
+	change := &git.GitChange{
+		ChangeType: &git.VersionControlChangeTypeValues.Delete,
+		Item: git.GitItem{
+			Path: &file,
+		},
+	}
+
+	_, err = clients.GitReposClient.CreatePush(ctx, git.CreatePushArgs{
+		RepositoryId: &repo,
+		Push: &git.GitPush{
+			RefUpdates: &[]git.GitRefUpdate{
+				{
+					Name:        &branch,
+					OldObjectId: &objectID,
+				},
+			},
+			Commits: &[]git.GitCommitRef{
+				{
+					Comment: &message,
+					Changes: &[]interface{}{change},
+				},
+			},
+		},
+	})
+	if err != nil {
+		return nil
+	}
+
+	return nil
+}
+
+// checkRepositoryBranchExists tests if a branch exists in a repository.
+func checkRepositoryBranchExists(c *client.AggregatedClient, repo, branch string) error {
+	branch = strings.TrimPrefix(branch, "refs/heads/")
+	ctx := context.Background()
+	_, err := c.GitReposClient.GetBranch(ctx, git.GetBranchArgs{
+		RepositoryId: &repo,
+		Name:         &branch,
+	})
+	return err
+}
+
+// checkRepositoryFileExists tests if a file exists in a repository.
+func checkRepositoryFileExists(c *client.AggregatedClient, repo, file, branch string) error {
+	ctx := context.Background()
+	_, err := c.GitReposClient.GetItem(ctx, git.GetItemArgs{
+		RepositoryId: &repo,
+		Path:         &file,
+		VersionDescriptor: &git.GitVersionDescriptor{
+			Version: &branch,
+		},
+	})
+	if err != nil {
+		return nil
+	}
+
+	return nil
+}
+
+func getLastCommitId(c *client.AggregatedClient, repo, branch string) (string, error) {
```

Comment:
The `branch` here  is in `ref/heads/xxxx` format,  but `getCommits` use `xxx` as the request parameter. [get commites](https://docs.microsoft.com/en-us/rest/api/azure/devops/git/commits/get%20commits?view=azure-devops-rest-6.0#on-a-branch).
![Uploading image.png…]()